File: deploy/www/verify/tests/duplicate.test.mjs
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { checkSignatureDuplicates } from '../verify.js';
import { parseHashSets } from '../parse.js';

test('clean signature list has no errors', () => {
  const items = [{ selector: 'sel1', alg: 'rsa-sha256' }, { selector: 'sel2', alg: 'ed25519-sha256' }];
  assert.deepEqual(checkSignatureDuplicates(items, 1), []);
});

test('duplicate selector is permerror', () => {
  // spec-06 §8.9: a Selector MUST NOT be present more than once
  const items = [{ selector: 'sel1', alg: 'rsa-sha256' }, { selector: 'sel1', alg: 'ed25519-sha256' }];
  assert.deepEqual(checkSignatureDuplicates(items, 3),
    ['PERMERROR DKIM2-Signature i=3 has a duplicate selector']);
});

test('duplicate selector is case-insensitive', () => {
  // Selector is a Domain (§3.5); DNS names are case-insensitive
  const items = [{ selector: 'Sel1', alg: 'rsa-sha256' }, { selector: 'sel1', alg: 'ed25519-sha256' }];
  const errs = checkSignatureDuplicates(items, 1);
  assert.equal(errs.length, 1);
  assert.match(errs[0], /has a duplicate selector/);
});

test('same algorithm twice with distinct selectors is allowed', () => {
  // spec-06 §8.9: one additional signature using the same algorithm MAY be
  // present provided a different Selector is used
  const items = [{ selector: 'sel1', alg: 'rsa-sha256' }, { selector: 'sel2', alg: 'rsa-sha256' }];
  assert.deepEqual(checkSignatureDuplicates(items, 1), []);
});

test('same algorithm three times has excess selectors', () => {
  const items = [
    { selector: 'sel1', alg: 'rsa-sha256' },
    { selector: 'sel2', alg: 'rsa-sha256' },
    { selector: 'sel3', alg: 'rsa-sha256' },
  ];
  assert.deepEqual(checkSignatureDuplicates(items, 2),
    ['PERMERROR DKIM2-Signature i=2 has more selectors than allowed']);
});

test('duplicate selector and excess-selector are independent', () => {
  // two sigs sharing an algorithm AND a selector is a duplicate-selector
  // error but NOT an excess-selector error (the count is 2, not 3+)
  const items = [{ selector: 'sel1', alg: 'rsa-sha256' }, { selector: 'sel1', alg: 'rsa-sha256' }];
  const errs = checkSignatureDuplicates(items, 1);
  assert.ok(errs.some((e) => e.includes('duplicate selector')));
  assert.ok(!errs.some((e) => e.includes('more selectors than allowed')));
});

test('parseHashSets keeps duplicate algorithms visible as a list', () => {
  // spec-06 §7.3: h= duplicate-algorithm detection must run over the parsed
  // LIST, never a deduplicated map (a map would silently overwrite the
  // second occurrence and make the duplicate undetectable).
  const sets = parseHashSets('sha256:AAA:BBB,sha256:CCC:DDD');
  assert.equal(sets.length, 2);
  assert.equal(sets[0].alg, 'sha256');
  assert.equal(sets[1].alg, 'sha256');
});

File: go/dkim2/duplicate_test.go
package dkim2

import "strings"
import "testing"

func TestCleanSignatureListHasNoErrors(t *testing.T) {
	errs := checkSignatureDuplicates([]SigItem{
		{Selector: "sel1", Algorithm: "rsa-sha256"},
		{Selector: "sel2", Algorithm: "ed25519-sha256"},
	}, 1)
	if len(errs) != 0 {
		t.Errorf("got %v, want no errors", errs)
	}
}

func TestDuplicateSelectorIsPermerror(t *testing.T) {
	errs := checkSignatureDuplicates([]SigItem{
		{Selector: "sel1", Algorithm: "rsa-sha256"},
		{Selector: "sel1", Algorithm: "ed25519-sha256"},
	}, 3)
	want := "PERMERROR DKIM2-Signature i=3 has a duplicate selector"
	if len(errs) != 1 || errs[0] != want {
		t.Errorf("got %v, want [%q]", errs, want)
	}
}

func TestDuplicateSelectorIsCaseInsensitive(t *testing.T) {
	errs := checkSignatureDuplicates([]SigItem{
		{Selector: "Sel1", Algorithm: "rsa-sha256"},
		{Selector: "sel1", Algorithm: "ed25519-sha256"},
	}, 1)
	if len(errs) != 1 || !strings.Contains(errs[0], "has a duplicate selector") {
		t.Errorf("got %v, want a duplicate-selector PERMERROR", errs)
	}
}

func TestSameAlgorithmTwiceAllowed(t *testing.T) {
	errs := checkSignatureDuplicates([]SigItem{
		{Selector: "sel1", Algorithm: "rsa-sha256"},
		{Selector: "sel2", Algorithm: "rsa-sha256"},
	}, 1)
	if len(errs) != 0 {
		t.Errorf("spec-06 §8.9 allows one additional same-algorithm signature with a distinct Selector; got %v", errs)
	}
}

func TestThreeSameAlgorithmHasExcessSelectors(t *testing.T) {
	errs := checkSignatureDuplicates([]SigItem{
		{Selector: "sel1", Algorithm: "rsa-sha256"},
		{Selector: "sel2", Algorithm: "rsa-sha256"},
		{Selector: "sel3", Algorithm: "rsa-sha256"},
	}, 2)
	if len(errs) != 1 || !strings.Contains(errs[0], "has more selectors than allowed") {
		t.Errorf("got %v, want an excess-selector PERMERROR", errs)
	}
}

func TestDuplicateSelectorAndExcessSelectorAreIndependent(t *testing.T) {
	// two sigs sharing an algorithm AND a Selector is a duplicate-selector
	// error but NOT an excess-selector error (the count is 2, not 3+)
	errs := checkSignatureDuplicates([]SigItem{
		{Selector: "sel1", Algorithm: "rsa-sha256"},
		{Selector: "sel1", Algorithm: "rsa-sha256"},
	}, 1)
	foundDup := false
	foundExcess := false
	for _, e := range errs {
		if strings.Contains(e, "duplicate selector") {
			foundDup = true
		}
		if strings.Contains(e, "more selectors than allowed") {
			foundExcess = true
		}
	}
	if !foundDup || foundExcess {
		t.Errorf("got %v, want only duplicate-selector error", errs)
	}
}

func TestDuplicateHashAlgorithmIsPermerror(t *testing.T) {
	_, err := parseMI("Message-Instance: m=4; h=sha256:AAA:BBB,sha256:CCC:DDD;")
	if err == nil || !strings.Contains(err.Error(), "has a duplicate hash algorithm") {
		t.Errorf("got %v, want a duplicate-hash-algorithm PERMERROR (spec-06 §7.3)", err)
	}
}
